File: languages/java/app-encryption/src/test/java/com/godaddy/asherah/crypto/keys/SharedCryptoKeyTest.java
package com.godaddy.asherah.crypto.keys;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.Instant;
import java.util.function.Consumer;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class SharedCryptoKeyTest {

  @Mock
  private CryptoKey sharedKeyMock;

  @InjectMocks
  private SharedCryptoKey sharedCryptoKey;

  @Test
  void testConstructor() {
    SharedCryptoKey sharedCryptoKey = new SharedCryptoKey(sharedKeyMock);
    assertNotNull(sharedCryptoKey);
  }

  @Test
  void testWithKeyFunction() {
    Function<byte[], ?> action = keyBytes -> null;

    sharedCryptoKey.withKey(action);
    verify(sharedKeyMock).withKey(action);
  }

  @Test
  void testWithKeyConsumer() {
    Consumer<byte[]> action = bytes -> {
    };

    sharedCryptoKey.withKey(action);
    verify(sharedKeyMock).withKey(action);
  }

  @Test
  void testGetCreated() {
    Instant expectedCreationTime = Instant.now();
    when(sharedKeyMock.getCreated()).thenReturn(expectedCreationTime);

    Instant actualCreationTime = sharedCryptoKey.getCreated();
    assertEquals(expectedCreationTime, actualCreationTime);
  }

  @Test
  void testMarkRevoked() {
    when(sharedKeyMock.isRevoked()).thenReturn(false);
    assertFalse(sharedCryptoKey.isRevoked());

    sharedCryptoKey.markRevoked();
    verify(sharedKeyMock).markRevoked();
  }

  @Test
  void testClose() {
    sharedCryptoKey.close();
    verify(sharedKeyMock, never()).close();
  }

}

File: csharp/AppEncryption/AppEncryption.Tests/Crypto/Keys/SharedCryptoKeyTest.cs
using System;
using GoDaddy.Asherah.Crypto.Keys;
using Moq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.Crypto.Keys
{
    [Collection("Logger Fixture collection")]
    public class SharedCryptoKeyTest
    {
        private readonly Mock<CryptoKey> sharedKeyMock;
        private readonly SharedCryptoKey sharedCryptoKey;

        public SharedCryptoKeyTest()
        {
            sharedKeyMock = new Mock<CryptoKey>();
            sharedCryptoKey = new SharedCryptoKey(sharedKeyMock.Object);
        }

        [Fact]
        public void TestConstructor()
        {
            SharedCryptoKey sharedCryptoKey = new SharedCryptoKey(sharedKeyMock.Object);
            Assert.NotNull(sharedCryptoKey);
        }

        [Fact]
        public void TestWithKeyFunction()
        {
            Func<byte[], object> action = keyBytes => null;
            sharedCryptoKey.WithKey(action);
            sharedKeyMock.Verify(x => x.WithKey(action), Times.Once);
        }

        [Fact]
        public void TestWithKeyAction()
        {
            Action<byte[]> action = bytes =>
            {
            };

            sharedCryptoKey.WithKey(action);
            sharedKeyMock.Verify(x => x.WithKey(action), Times.Once);
        }

        [Fact]
        public void TestGetCreated()
        {
            DateTimeOffset expectedCreationTime = DateTimeOffset.UtcNow;
            sharedKeyMock.Setup(x => x.GetCreated()).Returns(expectedCreationTime);
            DateTimeOffset actualCreatedTime = sharedCryptoKey.GetCreated();
            Assert.Equal(expectedCreationTime, actualCreatedTime);
        }

        [Fact]
        public void TestMarkRevoked()
        {
            sharedKeyMock.Setup(x => x.IsRevoked()).Returns(false);
            Assert.False(sharedCryptoKey.IsRevoked());
            sharedCryptoKey.MarkRevoked();
            sharedKeyMock.Verify(x => x.MarkRevoked(), Times.Once);
        }

        [Fact]
        public void TestDispose()
        {
            sharedCryptoKey.Dispose();
            sharedKeyMock.Verify(x => x.Dispose(), Times.Never);
        }
    }
}
